File: LeetCodes/island-perimeter/5-25-2022, 5_24_25 PM/island-perimeter.cpp
// Link: https://leetcode.com/problems/island-perimeter

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: island-perimeter
 *  RUNTIME: 178 ms
 *  MEMORY: 96.6 MB
 *  DATE: 5-25-2022, 5:24:25 PM
 *
 */

// Solution:

#define f(i, n) for (int i = 0; i < (n); i++)
class Solution {
public:
int dfs(int i, int j, vector<vector<int>> &grid)
{
    int perimeter = 0;

    if (i < 0 || j < 0)
        return 1;
    if (i >= grid.size() || j >= grid[0].size())
        return 1;
    if (grid[i][j] == 0)
        return 1;
    if (grid[i][j] == 1)
    {
        grid[i][j] = -1;
        perimeter += dfs(i - 1, j, grid);
        perimeter += dfs(i + 1, j, grid);
        perimeter += dfs(i, j - 1, grid);
        perimeter += dfs(i, j + 1, grid);
    }
    return perimeter;
}
    
    int islandPerimeter(vector<vector<int>> &grid)
{
    vector<vector<bool>> vis;
    int n = grid.size();
    int m = grid[0].size();
    int ans = 0;

    f(i, n)
        f(j, m) if (grid[i][j] == 1)
            ans += dfs(i, j, grid);
    return ans;
}

};

File: LeetCodes/island-perimeter/4-18-2024, 5_33_18 PM/island-perimeter.java
// Link: https://leetcode.com/problems/island-perimeter

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: island-perimeter
 *  RUNTIME: 5 ms
 *  MEMORY: 45.7 MB
 *  DATE: 4-18-2024, 5:33:18 PM
 *
 */

// Solution:

class Solution {
    public int islandPerimeter(int[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        int[] dir = {-1, 0, 1, 0, -1};
        int count = 0;

        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                if(grid[i][j] == 1) {
                    for(int k=0; k<4; k++) {
                        int x = i + dir[k];
                        int y = j + dir[k+1];
                        if(x >= n || x < 0 || y >= m || y < 0 || grid[x][y] == 0) {
                            count++;
                        }
                    }
                }
            }
        }
        return count;
    }
}
